src/Features/Core/Portable/EditAndContinue/DebuggingSession.cs

```diff
@@ -1,20 +1,327 @@
+                }
+
+                if (!update.ModuleReaders.IsEmpty)
+                {
+                    if (_lazyBaselineModuleReaders == null)
+                    {
+                        _lazyBaselineModuleReaders = new List<IDisposable>();
+                    }
+
+                    _lazyBaselineModuleReaders.AddRange(update.ModuleReaders);
+                }
+            }
+
+            // add to module updates map:
+            lock (_moduleUpdatesGuard)
+            {
+                void addDeltas(ImmutableArray<Deltas> deltas)
+                {
+                    foreach (var delta in deltas)
+                    {
+                        if (!_moduleUpdates.TryGetValue(delta.Mvid, out var existing))
+                        {
+                            _moduleUpdates.Add(delta.Mvid, existing = new List<Deltas>());
+                        }
+
+                        existing.Add(delta);
+                    }
+                }
+
+                addDeltas(update.Deltas);
+                addDeltas(update.DeferredDeltas);
+            }
+
+            LastCommittedSolution = update.Solution;
+        }
+
+        /// <summary>
+        /// Reads the MVID of a compiled project.
+        /// </summary>
+        /// <returns>
+        /// An MVID and an error message to report, in case an IO exception occurred while reading the binary.
+        /// The MVID is default if either project not built, or an it can't be read from the module binary.
+        /// </returns>
+        public async Task<(Guid Mvid, Diagnostic Error)> GetProjectModuleIdAsync(Project project, CancellationToken cancellationToken)
+        {
+            lock (_projectModuleIdsGuard)
+            {
+                if (_projectModuleIds.TryGetValue(project.Id, out var id))
+                {
+                    return id;
+                }
+            }
+
+            (Guid Mvid, Diagnostic Error) readMvid()
+            {
+                // TODO: use intermediate path
+                var filePath = project.OutputFilePath;
+
+                Contract.ThrowIfFalse(PathUtilities.IsAbsolute(filePath));
+
+                try
+                {
+                    return (ReadModuleVersionId(FileUtilities.OpenRead(filePath)), Error: null);
+                }
+                catch (Exception e) when (e is FileNotFoundException || e is DirectoryNotFoundException)
+                {
+                    return (Mvid: default, Error: null);
+                }
+                catch (Exception e)
+                {
+                    var descriptor = EditAndContinueDiagnosticDescriptors.GetDescriptor(EditAndContinueErrorCode.ErrorReadingFile);
+                    return (Mvid: default, Error: Diagnostic.Create(descriptor, Location.None, new[] { filePath, e.Message }));
+                }
+            }
+
+            var newId = await Task.Run(readMvid, cancellationToken).ConfigureAwait(false);
```

Comment:
❕  This method is already asynchronous. It should not need to use `Task.Run`. Ideally what I expect to see here is a direct call to `readMvid()`. For anything else, we should discuss this further.